### PR TITLE
Fixed #35111 -- Fixed compilation of DateField __in/__range rhs on SQLite and MySQL.

### PR DESCRIPTION
**Pull request by charettes**
ticket-35111

### CODE REVIEW

File: tests/expressions/tests.py [hunk 1]
```diff
@@ -1260,6 +1260,41 @@ def test_in_lookup_allows_F_expressions_and_expressions_for_datetimes(self):
         queryset = Result.objects.filter(result_time__range=within_experiment_time)
         self.assertSequenceEqual(queryset, [r1])
 
+    def test_range_lookup_allows_F_expressions_and_expressions_for_dates(self):
```
Comment (charettes) [inline tests/expressions/tests.py:1263]: These tests are quite heavy but I figured style coherency was worth more than optimization here.

File: django/db/models/lookups.py [hunk 2]
```diff
@@ -268,11 +268,16 @@ def get_db_prep_lookup(self, value, connection):
...
+        if not self.get_db_prep_lookup_value_is_iterable:
+            value = [value]
         return (
             "%s",
-            [get_db_prep_value(v, connection, prepared=True) for v in value]
-            if self.get_db_prep_lookup_value_is_iterable
-            else [get_db_prep_value(value, connection, prepared=True)],
+            [
+                v
+                if hasattr(v, "as_sql")
```
Comment (charettes) [inline django/db/models/lookups.py:277]: `hasattr(v, "as_sql")` is coherent with the other _is direct value_ checks.

**Comment by charettes** [comment]
@felixxm I think it is yes!

File: tests/expressions/tests.py (hunk 1 above)
Comment (felixxm) [reply tests/expressions/tests.py]: I think it's fine to use `subTest()` here.

Comment (charettes) [reply tests/expressions/tests.py]: Should you think it's worth adapting `test_in_lookup_allows_F_expressions_and_expressions_for_integers` as well to test different output fields under a single `test_in_lookup_allows_F_expressions_and_expressions` test with multiple `subTest`?

Comment (felixxm) [reply tests/expressions/tests.py]: Probably not worth changing, let's keep a data type in the test method name. Should be fine to keep datetimes and dates under `_for_dates` :thinking:

Comment (felixxm) [reply tests/expressions/tests.py:1228]: What do you think?

Comment (charettes) [reply tests/expressions/tests.py:1228]: That looks great 👍

Comment (felixxm) [reply tests/expressions/tests.py:1228]: :handshake: :tada:

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Pull request merged**